### PR TITLE
Add sdk.Context and vm.EVM to precompiles

### PR DESCRIPTION
**Pull request by loredanacirstea**
## Description

Add context to precompiles: both Cosmos `sdk.Context` and `vm.EVM`.
vm.EVM provides block, tx, along with state access.
And if this PR https://github.com/ethereum/go-ethereum/pull/26119 is merged, it will also provide current call/subcall context.

Related: https://github.com/evmos/ethermint/pull/1131

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Review by a1love-web — APPROVED** [review]

File: x/evm/vm/geth/geth.go [hunk 1]
```diff
@@ -55,20 +69,24 @@ func (e EVM) Config() vm.Config {
...
+	p, found = e.precompiles[addr]
 	return p, found
 }
 
 // ActivePrecompiles returns a list of all the active precompiled contract addresses
 // for the current chain configuration.
-func (EVM) ActivePrecompiles(rules params.Rules) []common.Address {
-	return vm.ActivePrecompiles(rules)
+func (e EVM) ActivePrecompiles(rules params.Rules) []common.Address {
+	// TODO e.precompiles
```
Comment (loredanacirstea) [reply x/evm/vm/geth/geth.go:79]: Yes, this `ActivePrecompiles` is not actually used now. I changed it when I was testing and forgot to revert the changes. For clarity, I will remove them.

File: x/evm/keeper/keeper.go [hunk 2]
```diff
@@ -56,8 +56,8 @@ type Keeper struct {
 	// EVM Hooks for tx post-processing
 	hooks types.EvmHooks
 
-	// custom stateless precompiled smart contracts
-	customPrecompiles evm.PrecompiledContracts
+	// Precompile extensions
+	getPrecompilesExtended func(ctx sdk.Context, evm *vm.EVM) evm.PrecompiledContracts
```
Comment (itsdevbear) [inline x/evm/keeper/keeper.go:60]: Could make this function take an address and return a singular precompile. Opens up more modularity.

File: x/evm/vm/geth/geth.go (hunk 1 above)
Comment (itsdevbear) [inline x/evm/vm/geth/geth.go:79]: This requires a geth fork right?

Comment (loredanacirstea) [reply x/evm/vm/geth/geth.go]: Ethermint uses its own `ActivePrecompiles` function in `state_transition.go`. The only other places I've seen `vm.ActivePrecompiles` used in go-ethereum is tracer-related.

**Comment by fedekunze** [comment]
marking this as Draft until we have more clarity on the final implementation

**Pull request closed without merging**